File: lib/services/analytics.dart
import 'dart:async';

import 'package:aptabase_flutter/aptabase_flutter.dart';
import 'package:flutter/foundation.dart';
import 'package:go_router/go_router.dart';

/// Privacy-first, cookieless usage analytics (Aptabase).
///
/// Why this needs no cookie / consent banner: Aptabase sets no cookies, stores
/// no persistent user identifier (session ids are generated in memory and
/// expire after inactivity — they are never written to disk), and collects no
/// personal data. The server derives only a coarse country from the request IP
/// and then discards it. The only thing kept on-device is a short-lived outbox
/// of not-yet-sent events (IndexedDB on web), which is deleted once delivered.
///
/// The project's own Aptabase (EU) ingestion key is baked in as the default, so
/// analytics works on every platform (web + mobile + desktop) with no per-build
/// flags. App keys are public client-side write keys — they ship inside the web
/// bundle and app binary regardless — so this is not a secret. Aptabase tags
/// debug-build events (`isDebug`) and the dashboard keeps Debug and Release data
/// separate, so local dev runs never pollute production stats.
///
/// Override or disable per build:
///   --dart-define=APTABASE_APP_KEY=A-EU-0000000000   (use a different key)
///   --dart-define=APTABASE_APP_KEY=                   (empty == analytics off)
/// For a self-hosted instance the key's region part is `SH`; also pass:
///   --dart-define=APTABASE_HOST=https://analytics.example.com
/// When the key is empty/garbled, every method here is a no-op: nothing is
/// collected, no plugin is initialised, and no network call is made.
class Analytics {
  const Analytics._();

  static const String _appKey = String.fromEnvironment(
    'APTABASE_APP_KEY',
    defaultValue: 'A-EU-2341128762',
  );
  static const String _host = String.fromEnvironment('APTABASE_HOST');

  /// A key is usable only if it has the `A-REG-0000000000` shape that
  /// [Aptabase.init] asserts on. We pre-check it so a missing/garbled key
  /// leaves analytics cleanly disabled instead of crashing startup in debug.
  static bool get _hasKey => _appKey.split('-').length == 3;

  /// Set once [init] has actually started the SDK; gates every send so we never
  /// touch Aptabase's `late` fields before it was initialised.
  static bool _started = false;

  /// Last route reported, used to drop the duplicate notifications go_router
  /// emits for a single navigation.
  static String? _lastRoute;

  /// Starts the SDK when a key is configured; otherwise a no-op.
  static Future<void> init() async {
    if (!_hasKey) return;
    try {
      await Aptabase.init(
        _appKey,
        _host.isEmpty ? const InitOptions() : InitOptions(host: _host),
      );
      _started = true;
    } catch (e) {
      // Analytics must never be able to break app startup.
      debugPrint('Analytics init failed: $e');
    }
  }

  /// Fire-and-forget custom event. Safe to call whether or not analytics ran.
  static void track(String event, [Map<String, dynamic>? props]) {
    if (!_started) return;
    // Not awaited: trackEvent only queues locally and flushes in the
    // background, so callers never wait on the network.
    unawaited(Aptabase.instance.trackEvent(event, props));
  }

  /// Subscribes to [router] so each navigation records a `screen_view`. No-op
  /// when analytics is disabled, so the listener isn't even attached then.
  static void attachRouter(GoRouter router) {
    if (!_started) return;
    void report() {
      try {
        _reportRoute(router.routerDelegate.currentConfiguration.uri.path);
      } catch (_) {
        // The first route may not be resolved yet; the listener below will
        // pick it up on the initial notification.
      }
    }

    router.routerDelegate.addListener(report);
    report();
  }

  static void _reportRoute(String path) {
    if (path.isEmpty || path == _lastRoute) return;
    _lastRoute = path;
    track('screen_view', {'route': path});
  }
}

File: src/lib/services/analytics.ts
// Privacy-first, cookieless usage analytics (Aptabase).
//
// Ported from lib/services/analytics.dart, but without the SDK: Aptabase's
// ingestion is one POST of a JSON envelope, and the Flutter plugin's extra —
// an on-device outbox — costs more than the events are worth here. An event
// that fails to send is simply dropped.
//
// Why this needs no cookie or consent banner: no cookie is set, no persistent
// identifier is stored (the session id lives in memory and expires after
// inactivity), and no personal data is collected. The server derives a coarse
// country from the request IP and then discards it.
//
// The ingestion key is a public client-side write key — it ships inside any
// web bundle regardless — so it is not a secret. Override or disable it per
// build with PUBLIC_APTABASE_APP_KEY (empty = analytics off entirely).

import { browser, dev } from '$app/environment';
import { env } from '$env/dynamic/public';

const DEFAULT_APP_KEY = 'A-EU-2341128762';

/** Ingestion host per key region, as the Aptabase SDKs resolve them. */
const HOSTS: Record<string, string> = {
	EU: 'https://eu.aptabase.com',
	US: 'https://us.aptabase.com',
	DEV: 'https://localhost:3000'
};

const appKey = env.PUBLIC_APTABASE_APP_KEY ?? DEFAULT_APP_KEY;
const region = appKey.split('-')[1] ?? '';
/** A self-hosted key (region `SH`) needs its host given explicitly. */
const host = env.PUBLIC_APTABASE_HOST || HOSTS[region] || '';

/** Only a well-formed `A-REG-0000000000` key with a known host can send. */
const enabled = appKey.split('-').length === 3 && host !== '';

/** A session is a run of activity; after this much idle time, a new one starts. */
const SESSION_TIMEOUT_MS = 60 * 60 * 1000;

let sessionId = '';
let lastTouched = 0;

/** In-memory only — nothing about the session is ever written to disk. */
function currentSession(): string {
	const now = Date.now();
	if (!sessionId || now - lastTouched > SESSION_TIMEOUT_MS) {
		sessionId = `${now}${Math.floor(Math.random() * 1e8)}`;
	}
	lastTouched = now;
	return sessionId;
}

/**
 * Fire-and-forget custom event. Safe to call from anywhere, including during
 * prerendering, where it is a no-op — analytics must never be able to break a
 * page, so every failure is swallowed.
 */
export function track(name: string, props: Record<string, string | number | boolean> = {}): void {
	if (!browser || !enabled) return;
	void fetch(`${host}/api/v0/event`, {
		method: 'POST',
		headers: { 'Content-Type': 'application/json', 'App-Key': appKey },
		// The request must not hold a navigation open.
		keepalive: true,
		body: JSON.stringify({
			timestamp: new Date().toISOString(),
			sessionId: currentSession(),
			eventName: name,
			systemProps: {
				isDebug: dev,
				locale: navigator.language,
				appVersion: '',
				sdkVersion: 'aptabase-web-fetch@1.0.0'
			},
			props
		})
	}).catch(() => {
		// Dropped. An event is never worth a console error in a learner's browser.
	});
}

let lastRoute: string | null = null;

/** Records one `screen_view`, ignoring the repeat notifications a nav emits. */
export function trackScreenView(path: string): void {
	if (!path || path === lastRoute) return;
	lastRoute = path;
	track('screen_view', { route: path });
}
